komga/src/main/kotlin/org/gotson/komga/interfaces/rest/SeriesController.kt

```diff
@@ -323,57 +324,77 @@ class SeriesController(
     if (!principal.user.canAccessLibrary(series)) throw ResponseStatusException(HttpStatus.FORBIDDEN)
 
-    val thumbnail = seriesLifecycle.getThumbnailBytes(seriesId, principal.user.id)
+    val thumbnail = if (thumbnailId != null) {
+      seriesLifecycle.getThumbnailBytesById(seriesId, thumbnailId)
+    } else {
+      seriesLifecycle.getThumbnailBytes(seriesId, principal.user.id)
+    }
     return thumbnail ?: throw ResponseStatusException(HttpStatus.NOT_FOUND)
   }
 
+  @GetMapping(value = ["{seriesId}/thumbnails"], produces = [MediaType.APPLICATION_JSON_VALUE])
+  fun getSeriesThumbnails(
+    @AuthenticationPrincipal principal: KomgaPrincipal,
+    @PathVariable(name = "seriesId") seriesId: String
+  ): Collection<SeriesThumbnailDto> {
+    seriesRepository.findByIdOrNull(seriesId) ?: throw ResponseStatusException(HttpStatus.NOT_FOUND)
+    if (!principal.user.canAccessLibrary(seriesId)) throw ResponseStatusException(HttpStatus.FORBIDDEN)
+
+    return seriesLifecycle.getAllThumbnailsForSeries(seriesId)
+      .map {
+        SeriesThumbnailDto(
+          id = it.id,
+          seriesId = it.seriesId,
+          type = it.type,
+          selected = it.selected
+        )
+      }
+  }
+
   @PatchMapping(value = ["{seriesId}/thumbnail"], consumes = [MediaType.MULTIPART_FORM_DATA_VALUE])
   @PreAuthorize("hasRole('$ROLE_ADMIN')")
   @ResponseStatus(HttpStatus.ACCEPTED)
-  fun patchCustomSeriesThumbnail(
-    @AuthenticationPrincipal principal: KomgaPrincipal,
+  fun patchUserUploadedSeriesThumbnail(
     @PathVariable(name = "seriesId") seriesId: String,
     @RequestParam("image") image: MultipartFile
   ) {
     val series = seriesRepository.findByIdOrNull(seriesId) ?: throw ResponseStatusException(HttpStatus.NOT_FOUND)
-    try {
-      ImageIO.read(image.inputStream).toString()
-      seriesLifecycle.addCustomThumbnailForSeries(series, image.bytes)
-    } catch (e: Exception) {
-      when (e) {
-        is IOException,
-        is IllegalArgumentException,
-        is NullPointerException -> throw ResponseStatusException(HttpStatus.UNSUPPORTED_MEDIA_TYPE)
-        is DataAccessException -> throw ResponseStatusException(HttpStatus.SERVICE_UNAVAILABLE)
-        else -> throw ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR)
-      }
+    val type = contentDetector.detectMediaType(BufferedInputStream(image.inputStream))
+    if (!contentDetector.isImage(type)) {
+      throw ResponseStatusException(HttpStatus.UNSUPPORTED_MEDIA_TYPE)
     }
+    seriesLifecycle.addUserUploadedThumbnailForSeries(series, image.bytes)
   }
 
-  @DeleteMapping("{seriesId}/thumbnail")
+  @PatchMapping("{seriesId}/thumbnail/{thumbnailId}")
   @PreAuthorize("hasRole('$ROLE_ADMIN')")
   @ResponseStatus(HttpStatus.ACCEPTED)
-  fun deleteCustomSeriesThumbnail(
-    @AuthenticationPrincipal principal: KomgaPrincipal,
+  fun patchMarkSelectedUserUploadedSeriesThumbnail(
```

Comment:
we could theoretically mark any thumbnail as selected, not just user uploaded ones